File: part-02/src/test/java/eu/luminis/workshop/smallsteps/logic/domainservice/LegoStockReserveTest.java
package eu.luminis.workshop.smallsteps.logic.domainservice;

import eu.luminis.workshop.smallsteps.logic.domainservice.auth.AccessDeniedException;
import eu.luminis.workshop.smallsteps.logic.domainservice.helper.SetupLegoTestApp;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.LegoSetForBuilder;
import eu.luminis.workshop.smallsteps.logic.domainservice.state.StockState;
import eu.luminis.workshop.smallsteps.persistence.InMemoryLegoPartCatalog;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.assertj.core.api.Assertions.*;

public class LegoStockReserveTest {
    SetupLegoTestApp app;

    @BeforeEach
    void setUp() {
        app = new SetupLegoTestApp();
    }

    @Test
    public void should_reserve_lego_set_when_in_stock() {
        StockState initialState = new StockState(
                Map.of(app.millenniumFalcon, 1),
                null,
                null,
                null,
                null
        );
        LegoStock legoStock = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        StockState newState = legoStock.reserve(app.millenniumFalcon);

        assertThat(newState.getAvailableStock()).isNotEmpty().contains(entry(app.millenniumFalcon, 0));
        assertThat(newState.getReserved()).isNotEmpty().contains(new LegoSetForBuilder(app.millenniumFalcon, app.harry));
        assertThat(newState.getAtBuilder()).isEqualTo(initialState.getAtBuilder());
        assertThat(newState.getIncompleteStock()).isEqualTo(initialState.getIncompleteStock());
        assertThat(newState.getIncompleteReturnHistory()).isEqualTo(initialState.getIncompleteReturnHistory());
    }

    @Test
    public void should_fail_to_reserve_when_lego_set_not_in_store() {
        LegoStock legoStock = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), new StockState());

        assertThatThrownBy(() -> {
            legoStock.reserve(app.millenniumFalcon);
        }).isInstanceOf(IllegalArgumentException.class)
                .hasMessage("LegoSet " + app.millenniumFalcon.getNumber() + " is not in stock");
    }

    @Test
    public void should_fail_to_reserve_when_lego_set_0_in_store() {
        StockState initialState = new StockState(
                Map.of(app.millenniumFalcon, 0),
                null,
                null,
                null,
                null
        );
        LegoStock legoStock = new LegoStock(app.harryAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            legoStock.reserve(app.millenniumFalcon);
        }).isInstanceOf(IllegalArgumentException.class)
                .hasMessage("LegoSet " + app.millenniumFalcon.getNumber() + " is not in stock");
    }

    @Test
    public void should_fail_to_reserve_when_not_authenticated_as_builder() {
        StockState initialState = new StockState(
                Map.of(app.millenniumFalcon, 1),
                null,
                null,
                null,
                null
        );
        LegoStock legoStock = new LegoStock(app.bussumAuth, new InMemoryLegoPartCatalog(), initialState);

        assertThatThrownBy(() -> {
            legoStock.reserve(app.millenniumFalcon);
        }).isInstanceOf(AccessDeniedException.class);

    }
}

File: part-02/src/test/kotlin/eu/luminis/workshop/smallsteps/logic/domainService/LegoStock - ReserveTest.kt
package eu.luminis.workshop.smallsteps.logic.domainService

import eu.luminis.workshop.smallsteps.logic.domainService.auth.AccessDeniedException
import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper
import eu.luminis.workshop.smallsteps.logic.domainService.helper.LegoStockHelper.millenniumFalcon
import eu.luminis.workshop.smallsteps.logic.domainService.helper.TestUsers
import eu.luminis.workshop.smallsteps.logic.domainService.state.StockState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

@Suppress("ClassName")
internal class `LegoStock - ReserveTest` {
    @Test
    fun `should reserve lego set when in stock`() {
        // given
        val initialState = StockState(
            availableStock = mapOf(millenniumFalcon to 1),
            reserved = emptySet(),
        )
        val legoStock = LegoStockHelper.build(initialState)

        // when
        val newState = legoStock.reserve(millenniumFalcon)

        // then
        assertEquals(mapOf(millenniumFalcon to 0), newState.availableStock)
        assertEquals(setOf(millenniumFalcon to TestUsers.harry), newState.reserved)
        assertEquals(initialState.atBuilder, newState.atBuilder)
        assertEquals(initialState.incompleteStock, newState.incompleteStock)
        assertEquals(initialState.incompleteReturnHistory, newState.incompleteReturnHistory)
    }

    @Test
    fun `should fail to reserve when lego set not in store`() {
        // given
        val legoStock = LegoStockHelper.build(StockState())

        // then
        assertFailsWith<IllegalArgumentException> {
            // when
            legoStock.reserve(millenniumFalcon)
        }
    }

    @Test
    fun `should fail to reserve when lego set 0 in store`() {
        // given
        val initialState = StockState(
            availableStock = mapOf(millenniumFalcon to 0)
        )
        val legoStock = LegoStockHelper.build(initialState)

        // then
        assertFailsWith<IllegalArgumentException> {
            // when
            legoStock.reserve(millenniumFalcon)
        }
    }

    @Test
    fun `should fail to reserve when not authenticated as builder`() {
        // given
        val initialState = StockState(
            availableStock = mapOf(millenniumFalcon to 1),
            reserved = emptySet(),
        )
        val legoStock = LegoStockHelper.build(initialState, auth = TestUsers.bussumAuth)

        // then
        assertFailsWith<AccessDeniedException> {
            // when
            legoStock.reserve(millenniumFalcon)
        }
    }
}